logger/formatter.go

```diff
@@ -0,0 +1,193 @@
+package logger
+
+import (
+	"io"
+	"log"
+	"os"
+	"runtime"
+	"sync"
+	"time"
+
+	"github.com/mattn/go-isatty"
+)
+
+type (
+	formatter struct {
+		colored bool
+		prefix  string
+		out     io.Writer // destination for output
+		flag    int       // properties
+
+		mu  sync.Mutex // ensures atomic writes; protects the following fields
+		buf []byte     // for accumulating text to write
+	}
+)
+
+func newFormatter(out io.Writer) *formatter {
+	if isatty.IsTerminal(os.Stderr.Fd()) {
+		return &formatter{
+			out:     out,
+			colored: true,
+			flag:    log.Lshortfile,
+			buf:     make([]byte, 0, 120),
+		}
+	}
+	return &formatter{
+		out:     out,
+		colored: false,
+		prefix:  "go.d ",
+		flag:    log.Ldate | log.Ltime,
+		buf:     make([]byte, 0, 120),
+	}
+}
+
+func (l *formatter) SetOutput(out io.Writer) {
+	l.out = out
+}
+
+func (l *formatter) Output(severity Severity, module, job string, callDepth int, s string) {
+	now := time.Now() // get this early.
+	var file string
+	var line int
+	if l.flag&(log.Lshortfile|log.Llongfile) != 0 {
+		var ok bool
+		_, file, line, ok = runtime.Caller(callDepth)
+		if !ok {
+			file = "???"
+			line = 0
+		}
+	}
+
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	l.formatTimestamp(now)
+	l.buf = append(l.buf, l.prefix...)
+	l.formatSeverity(severity)
+	l.formatModuleJob(module, job)
+	l.formatFile(file, line)
+	l.buf = append(l.buf, s...)
+	if s == "" || s[len(s)-1] != '\n' {
+		l.buf = append(l.buf, '\n')
+	}
+	_, _ = l.out.Write(l.buf)
+	l.buf = l.buf[:0]
+}
+
+func (l *formatter) formatModuleJob(module string, job string) {
+	l.buf = append(l.buf, module...)
+	l.buf = append(l.buf, '[')
+	l.buf = append(l.buf, job...)
+	l.buf = append(l.buf, "] "...)
+}
+
+// formatTimestamp writes timestamp to buf
+func (l *formatter) formatTimestamp(t time.Time) {
+	if l.flag&(log.Ldate|log.Ltime|log.Lmicroseconds) != 0 {
+		if l.flag&log.LUTC != 0 {
+			t = t.UTC()
+		}
+		if l.flag&log.Ldate != 0 {
+			year, month, day := t.Date()
+			itoa(&l.buf, year, 4)
+			l.buf = append(l.buf, '-')
+			itoa(&l.buf, int(month), 2)
+			l.buf = append(l.buf, '-')
+			itoa(&l.buf, day, 2)
+			l.buf = append(l.buf, ' ')
+		}
+		if l.flag&(log.Ltime|log.Lmicroseconds) != 0 {
+			hour, min, sec := t.Clock()
+			itoa(&l.buf, hour, 2)
+			l.buf = append(l.buf, ':')
+			itoa(&l.buf, min, 2)
+			l.buf = append(l.buf, ':')
+			itoa(&l.buf, sec, 2)
+			if l.flag&log.Lmicroseconds != 0 {
+				l.buf = append(l.buf, '.')
+				itoa(&l.buf, t.Nanosecond()/1e3, 6)
+			}
+			l.buf = append(l.buf, ' ')
+		}
+		l.buf[len(l.buf)-1] = ':'
+		l.buf = append(l.buf, ' ')
+	}
+}
+
+func (l *formatter) formatSeverity(severity Severity) {
+	if l.colored {
+		switch severity {
+		case DEBUG:
+			l.buf = append(l.buf, "\x1b[0;36m[ "...)
+		case INFO:
+			l.buf = append(l.buf, "\x1b[0;32m[ "...)
+		case WARNING:
+			l.buf = append(l.buf, "\x1b[0;33m[ "...)
```

Comment:
Hey, can we have human readable colors?

Reply:
<img width="998" alt="2018-12-04 15 13 24" src="https://user-images.githubusercontent.com/8895721/49422445-3d1adc80-f7d7-11e8-96dc-eb3b3aca6cb2.png">

Reply:
i mean in the code

Reply:
It's highly optimized code. 3x-5x faster than previous code.

Reply:
simple comments will do :smile:

Reply:
I see 2 spaces after `INFO` and `WARN`

Reply:
please fix it

Reply:
we need the space to align the log.

Reply:
oh, thats on purpose, i see

Reply:
I wonder how it looks with `CRITICAL`

Reply:
i think it's just `CRIT`, ok

Reply:
in CLI, `CRITICAL` => `[ CRIT ]`